### PR TITLE
feat: push all possible filters down to parquet exec

### LINKED ISSUE #992: Row group pruning predicate does not coerce data types.
### What type of bug is this?

Unexpected error

### What subsystems are affected?

Datanode

### What happened?

When passing time range predicate with a time unit differs from timestamp column time unit to prune row groups, it will fail to prune row groups since built binary expression does not coerce time units (arrow considers timestamp with different unit as different data types).

### What operating system did you use?

macOS Montery

### Relevant log output and stack trace

```bash
2023-02-14T08:47:31.317143Z ERROR table::predicate: Failed to create predicate for expr, error: Internal("The type of Timestamp(Millisecond, None) >= Timestamp(Second, None) of binary physical should be same")
```
```


### How can we reproduce the bug?

```bash
1. Create a table with ts column in milliseconds;
2. Create a query condition on ts column with time unit seconds;
3. Using that query condition to prune row groups and it will report an error.
```

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This PR pushes all possible filters down to Parquet exec to improve scan efficiency. Also this PR coerces time range predicate data types to timestamp type in storage schemas to address #992

## Future work
- PagePruningPredicate

## Checklist

- [X]  I have written the necessary rustdoc comments.
- [X]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
Fixes #992

### CODE REVIEW

File: src/servers/src/mysql/handler.rs [hunk 1]
```diff
@@ -206,7 +206,7 @@ impl<W: AsyncWrite + Send + Sync + Unpin> AsyncMysqlShim<W> for MysqlInstanceShi
         query: &'a str,
         w: StatementMetaWriter<'a, W>,
     ) -> Result<()> {
-        let raw_query = query.clone();
+        let raw_query = <&str>::clone(&query);
```
Comment (v0y4g3r) [inline src/servers/src/mysql/handler.rs:209]: Curiously clippy complains about this line for "cloning a double reference".

File: src/table/src/predicate.rs [hunk 2]
```diff
@@ -12,66 +12,92 @@
...
+    /// Returns error when failed to converting exprs.
+    pub fn try_new(exprs: Vec<Expr>, schema: SchemaRef) -> error::Result<Self> {
+        let arrow_schema = schema.arrow_schema();
+        let df_schema = arrow_schema
+            .clone()
+            .to_dfschema_ref()
+            .context(error::DatafusionSnafu)?;
+
+        // `execution_props` provides variables required by evaluation.
+        let execution_props = &ExecutionProps::new();
```
Comment (killme2008) [inline src/table/src/predicate.rs:58]: I think `execution_props` should be retrieved from the query engine to respect the query engine settings.

https://docs.rs/datafusion/latest/datafusion/execution/context/struct.SessionState.html#method.execution_props

https://github.com/GreptimeTeam/greptimedb/blob/f287d3115b5ea38f72c9267d8437ecb6067ff57c/src/query/src/query_engine/state.rs#L57

File: src/storage/src/sst/pruning.rs [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+            projection_mask: projection_mask.clone(),
+            physical_expr: expr.clone(),
+        }) as _);
+    }
+
+    Ok(datafusion_predicates)
+}
+
+#[derive(Debug)]
+struct DatafusionArrowPredicate {
```
Comment (killme2008) [inline src/storage/src/sst/pruning.rs:107]: Adds some tests for DatafusionArrowPredicate

**Review by killme2008 — COMMENTED** [review]
Great work!

File: src/storage/src/chunk.rs [hunk 4]
```diff
@@ -270,7 +276,12 @@ impl ChunkReaderBuilder {
     /// Build time range predicate from schema and filters.
     pub fn build_time_range_predicate(&self) -> TimestampRange {
         let Some(ts_col) = self.schema.user_schema().timestamp_column() else { return TimestampRange::min_to_max() };
-        TimeRangePredicateBuilder::new(&ts_col.name, &self.filters).build()
+        let unit = ts_col
+            .data_type
+            .as_timestamp()
+            .expect("Timestamp column must have timestamp-compatible type")
```
Comment (evenyag) [inline src/storage/src/chunk.rs:282]: question: What if the data type of the timestamp column is Int64?

File: src/table/src/predicate.rs (hunk 2 above)
Comment (v0y4g3r) [reply src/table/src/predicate.rs:58]: I did not use `ExectionProps` from `SesstionState` in that:
- We cannot pass the session state down to the table scan stage.
- The `SessionState::execution_props` field is [initialize to `ExecutionProps::new()`](https://docs.rs/datafusion/25.0.0/src/datafusion/execution/context.rs.html#1488) and we never register any variable provider to `SessionState`, thus it's the same as `ExecutionProps::new()` here.

Comment (killme2008) [reply src/table/src/predicate.rs:58]: Yep, but when we start to register some variables to the session state, we may forget here. 
Maybe we can add a TODO or a global variable in query crate.

File: src/storage/src/sst/parquet.rs [hunk 5]
```diff
@@ -277,26 +269,27 @@ impl ParquetReader {
...
         let mut builder = builder
-            .with_projection(projection)
+            .with_projection(projection_mask.clone())
             .with_row_groups(pruned_row_groups);
 
         // if time range row filter is present, we can push down the filter to reduce rows to scan.
-        if let Some(row_filter) =
-            build_time_range_row_filter(self.time_range, &store_schema, &parquet_schema_desc)
-        {
+        if let Some(row_filter) = build_row_filter(
```
Comment (evenyag) [inline src/storage/src/sst/parquet.rs:286]: question: Is the comment above still correct? Seems this function might return filters for other columns.

File: src/storage/src/sst/pruning.rs [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+use datatypes::prelude::ConcreteDataType;
+use parquet::arrow::arrow_reader::{ArrowPredicate, RowFilter};
+use parquet::arrow::ProjectionMask;
+use parquet::schema::types::SchemaDescriptor;
+use table::predicate::Predicate;
+
+use crate::error;
+use crate::schema::StoreSchema;
+
+/// Builds time range row filter.
```
Comment (evenyag) [inline src/storage/src/sst/pruning.rs:37]: question: Is this comment correct? Does the filter only apply the time range predicate?

File: src/storage/src/compaction/writer.rs [hunk 7]
```diff
@@ -31,53 +32,81 @@ pub(crate) async fn build_sst_reader(
...
-            left: ts_col,
-            op: Operator::Lt,
-            right: upper_bound_expr,
-        })),
-    });
-
-    Expr::from(expr)
+/// Build time range filter expr from lower (inclusive) and upper bound(exclusive).
+/// Returns `None` if time range overflows.
+fn build_time_range_filter(
```
Comment (evenyag) [inline src/storage/src/compaction/writer.rs:59]: question: If compaction uses this filter, will it returns rows out of the time range when the timestamp conversion overflows?

Also, converting ms to s might lose some precision.

File: src/storage/src/sst/pruning.rs (hunk 3 above)
Comment (evenyag) [inline src/storage/src/sst/pruning.rs:107]: Does datafusion provide similar struct so we can reuse it?

File: src/storage/src/chunk.rs (hunk 4 above)
Comment (v0y4g3r) [reply src/storage/src/chunk.rs:282]: > question: What if the data type of the timestamp column is Int64?

Int64 is treated as `TimestampMillisecond`.

https://github.com/GreptimeTeam/greptimedb/blob/30472cebaeee399ce163780ddb9714806c012ebf/src/datatypes/src/data_type.rs#L177

File: src/storage/src/sst/pruning.rs (hunk 3 above)
Comment (v0y4g3r) [reply src/storage/src/sst/pruning.rs:107]: datafusion does have a similar struct but it's private to `datafusion::physical_plan`.

File: src/storage/src/compaction/writer.rs (hunk 7 above)
Comment (v0y4g3r) [reply src/storage/src/compaction/writer.rs:59]: > question: If compaction uses this filter, will it returns rows out of the time range when the timestamp conversion overflows?
> 
> Also, converting ms to s might lose some precision.

For compaction, this timestamp range is always inferred from timestamp values in Parquet files, and then converted to `TimeUnit::Second`. For example, the original timestamp unit is nanosecond, converting any value with that time unit to `TimeUnit::Second` and then back to `TimeUnit::Nanosecond` won't cause an overflow.

**Review by evenyag — APPROVED** [review]
Mostly LGTM.

File: src/table/src/predicate.rs [hunk 8]
```diff
@@ -329,10 +368,15 @@ mod tests {
         (path, schema)
     }
 
-    async fn assert_prune(array_cnt: usize, predicate: Predicate, expect: Vec<bool>) {
+    async fn assert_prune(
+        array_cnt: usize,
+        filters: Vec<common_query::logical_plan::Expr>,
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/table/src/predicate.rs:373]: No, in the test mod only `datafusion_expr::Expr` is imported.

**Pull request merged**